File: python/Array_Hash/238. Product of Array Except Self.py
'''
Time: O(n)
Space: O(1)
'''

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [0] * len(nums)

        prefix = 1
        for i in range(len(nums)):
            res[i] = prefix
            prefix *= nums[i]

        postfix = 1
        for i in range(len(nums), -1, -1):
            res[i] *= postfix  # res[i] = prefix * postfix
            postfix *= nums[i]

        return res

File: javascript/Array_Hash/238. Product of Array Except Self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
    const res = new Array(nums.length).fill(1);

    let prefix = postfix = 1;
    let j = nums.length - 1;  // for postfix
    nums.forEach((val, i) => {
        // * prefix
        res[i] *= prefix;
        prefix *= val;
        // * postfix
        res[j] *= postfix;
        postfix *= nums[j];
        j--;
    });
    return res;
};